File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        sanitel = ListNode()
        sanitel.next = head
        #head is exclusive
        def reverse(head, k):
            if k == 1:
                return head.next
            p = head.next.next
            end = head.next
            while(k > 1):
                k = k - 1
                r = p.next
                end.next = p.next
                p.next = head.next
                head.next = p
                p = r
            return head.next
        #head is exclusive        
        def reverseK(head, k):
            m = k
            p = head
            while(m > 0):
                if not p:
                    return head.next
                p = p.next
                m = m - 1
            if not p:
                return head.next
            q = head.next
            head.next = reverse(head,k)
            q.next = reverseK(q,k)       
            return head.next
        return reverseK(sanitel,k)
                
            
        

File: 25-reverse-nodes-in-k-group/25-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dumb = new ListNode(0);
        dumb.next = head;
        ListNode p = head;
        int count = 1;
        ListNode res = new ListNode(0);
        ListNode resP = res;
        while(p != null){
            if(count != 0 && count%k == 0){
                //extract the k list
                ListNode nxt = p.next;
                p.next = null;
                ListNode newTail = dumb.next;
                resP.next = this.reverseList(newTail);
                dumb.next = nxt;
                resP = newTail;
                p = dumb.next;
            }else{
                p = p.next;
            }
            count++;
        }
        if(dumb.next != null){
            resP.next =dumb.next;
        }
        return res.next;
    }
    public ListNode reverseList(ListNode head){
        ListNode cur = head;
        ListNode prev = null;
        while(cur != null){
            ListNode nxt = cur.next;
            cur.next = prev;
            prev = cur;
            cur = nxt;
        }
        return prev;
        
        
        
        
        
        
    }
}